README
Sun Feb 22 09:08:47 1998, faith@cs.unc.edu


WHAT:

The files contained herein are designed to convert the data from the 
ftp://ftp.cs.unc.edu/pub/users/faith/dict/web1913-0.46-a.tar.gz
package into a working database and index for the DICT server.

The original data for this version of:
     Webster's Revised Unabridged Dictionary (G & C. Merriam Co., 1913,
     edited by Noah Porter). Online version prepared by MICRA, Inc.,
     Plainfield, N.J. and edited by Patrick Cassidy <cassidy@micra.com>.
     ftp://uiarchive.cso.uiuc.edu/pub/etext/gutenberg/etext96/pgw*
     http://humanities.uchicago.edu/forms_unrest/webster.form.html
is available from: ftp://ftp.uga.edu/pub/misc/webster/, but some patches
were applied to the files in the web1913-0.46-a.tar.gz distribution.

This package only contains conversion software, and contains NO DATA.  You
must obtain the data separately.  Please see the raw data for distribution
terms for the raw and formatted data.  The formatting _software_ is
currently distributed under the GNU General Public License.  If you need it
use it under another license, contact the author.


CONFIGURATION:

./configure options:
  --with-tmppath=PATH     use PATH for temporary files
  --with-datapath=PATH    use PATH for original database files
  --prefix=PREFIX         install architecture-independent files in PREFIX
  --datadir=DIR           read-only architecture-independent data in DIR

tmppath defaults to . and requires 40MB

datapath defaults to ./web1913-* and should contain the data found in
web1913-0.46-a.tar.gz [this software probably won't work with later
versions of the raw data, but you can try...]

prefix defaults to /usr

datadir defaults to /usr/lib, so the data files get put in /usr/lib/dict.
They are machine independent, you you'd want to change this to a shared
directory structure if you want the data shared between several servers.
For example, use --datadir=/usr/share to put the files in /usr/share/dict.

30 MB of data and 4MB of index will be written to ./web1913.dict and
./web1913.index -- you must have enough disk space available for these
files, for the original 42MB of raw files (in datapath), and for the
intermediate 40MB file (in tmppath).  This means you need close to 150MB of
free disk space to format this dataset.  Additionally, you need about 50MB
of swap space for the running programs.  Formatting takes less than 15
minutes on a typical Pentium 133MHz.

If you have libmaa installed on your system, then the local copy won't get
built unless you use --with-local-libmaa.  Otherwise, it will be built
automatically.

If you have the dictzip program installed on your system, then the final
30MB web1913.dict file will be compressed to about 10MB.  The file will be
readable with the GNU zip utilities.  If you don't want this compression
performed even though you have dictzip installed, make DICTZIP=cat and it
will be skipped.


EXECUTION:

make          builds the libraries and executables
make db       formats the database
make install  installs the database (no executables are installed)

When you're done with the "make db" step, it should say
      185399 headwords words for       115301 entries, total

[Versus 185774 headwords words for 115272 entries, total in 0.45.]
[Versus 187637 headwords words for 115283 entries, total in 0.42]
[Versus 202209 headwords words for 124121 entries, total in cide 0.43]

and you'll be using up about 80MB for everything (less if you have dictzip
installed and it automatically compressed the database).  Note that each of
the processes use about 40MB of memory, so you need a reasonable amount of
RAM or swap space.

PROBLEMS:

If the problem is with the data, you should contact MICRA via Patrick
Cassidy <cassidy@micra.com>.  Be nice and try to be helpful -- putting this
database together was a lot of work, and there are a lot of typos and
data-related problems.  If you want to edit definitions, you should
probably get the latest version of the data and work with that.  If you
want to redistribute the data, see the terms in the raw and formatted data
files (the terms are MICRA's, and are the same for the raw and formatted
data).

If the problem is with the programs, you should contact me, Rik Faith
(faith@cs.unc.edu).  Patches or suggestions are always welcome, but my time
is limited and I might not reply to your email immediately.  If its many
years after 1998, faith@acm.org might be a better alternative address.


ChangeLog
Sun Feb 22 19:29:21 1998  Rik Faith  <faith@rtp-nc4-08>

  * Makefile.in: Elide CIDE for now

  * scan.l, webfilter.l, Makefile.in, README, parse.y: More cross references
  Revert CIDE until more time is available to work on it

  * scan.l, README: Ignore <br/ -- it causes too many problems

  * parse.y, scan.l, webfilter.l, webfmt.c, Makefile.in, README:
  Update for 0.46 and CIDE

Mon Feb 16 00:05:03 1998  Rik Faith  <faith@rtp-nc4-08>

  * scan.l, webfilter.l, parse.y: Added cross references

Mon Jan 26 03:51:42 1998  Rik Faith  <faith@rtp-nc4-08>

  * webfilter.l, scan.l, README, parse.y, Makefile.in:
  Working on web1913 0.45 and {} cross references

Sat Jul 12 02:59:43 1997  Rik Faith  <faith@rtp-nc4-08>

  * scan.l, fmt.c, configure.in, README, Makefile.in:
  Minor changes preparing for distribution

Thu May 22 01:14:40 1997  Rik Faith  <faith@rtp-nc4-08>

  * Makefile.in, configure.in: Fixed raw data search strategy

  * Attic/iso-filter.dat, configure.in, README, Makefile.in:
  Preparing for release

Wed May 21 23:52:07 1997  Rik Faith  <faith@rtp-nc4-08>

  * webfmt.h, webfmt.c, webfilter.l, webcheck.l, scan.l, parse.y, fmt.c, configure.in, Makefile.in, Makefile.conf:
  Unified processor naming

  * Attic/iso-filter.dat: First attempt at ISO-8859-1 character use

  * scan.l, webfilter.l, fmt.h, fmt.c, configure.in, Makefile.in:
  Added copyright information from original data

Sat May 10 10:21:02 1997  Rik Faith  <faith@rtp-nc4-08>

  * scan.l, webfilter.l, parse.y, fmt.c, filter.dat:
  Updated routines for 0.42 database

Sun Mar 23 12:24:16 1997  Rik Faith  <faith@rtp-nc4-08>

  * webfilter.l, tex-like.dat, scan.l, webfmt.h, webfmt.c, fmt.h, parse.y, fmt.c, filter.dat:
  Finished preliminary formatting

Wed Mar 19 21:11:35 1997  Rik Faith  <faith@rtp-nc4-08>

  * webfilter.l, parse.y, scan.l, webfmt.c, Makefile.in, filter.dat:
  Finished up entities, now on to formatting

Tue Mar 18 22:33:00 1997  Rik Faith  <faith@rtp-nc4-08>

  * webcheck.l: MOre entities

  * scan.l, webcheck.l, Makefile.in: More entities...

  * scan.l, webfmt.h, parse.y, webfmt.c, fmt.h, fmt.c, webcheck.l, dump.h, configure.in, config.h.in, Makefile.in:
  Working toward nicely formatted output

Mon Mar 10 21:46:53 1997  Rik Faith  <faith@rtp-nc4-08>

  * configure.in, COPYING: More work

Sat Mar  8 22:36:16 1997  Rik Faith  <faith@rtp-nc4-08>

  * configure.in: Finish SunOS port

  * configure.in: Cleanup for SunOS

Fri Mar  7 20:37:09 1997  Rik Faith  <faith@rtp-nc4-08>

  * configure.in, config.sub: Reorganization for better packaging

Thu Sep 26 02:27:42 1996  Rik Faith  <faith@rtp-nc4-08>

  * install-sh, config.sub, configure.in, config.guess:
  General work on formatting and use with configure.

Tue Sep 24 01:07:41 1996  Rik Faith  <faith@rtp-nc4-08>

  * COPYING: Move to version 2.0

Mon Sep 23 15:28:28 1996  Rik Faith  <faith@rtp-nc4-08>

  * COPYING: dict-1.0



webfmt.c
/* webfmt.c -- Project Gutenberg Webster converter
 * Created: Sun Mar 16 09:29:39 1997 by faith@cs.unc.edu
 * Revised: Sun Feb 22 09:27:45 1998 by faith@acm.org
 * Copyright 1997, 1998 Rickard E. Faith (faith@acm.org)
 * 
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 1, or (at your option) any
 * later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 675 Mass Ave, Cambridge, MA 02139, USA.
 * 
 * $Id: webfmt.c,v 1.5 1998/02/22 15:14:58 faith Exp $
 * 
 */

#include "webfmt.h"
#include "parse.h"
#include <sys/utsname.h>

extern int        yy_flex_debug;
       int        database = 0;

static const char *id_string( const char *id )
{
   static char buffer[256];
   arg_List a = arg_argify( id, 0 );

   if (arg_count(a) >= 2)
      sprintf( buffer, "%s", arg_get( a, 2 ) );
   else
      buffer[0] = '\0';
   arg_destroy( a );
   return buffer;
}

static const char *pgw_get_banner( void )
{
   static char       *buffer= NULL;
   const char        *id = "$Id: webfmt.c,v 1.5 1998/02/22 15:14:58 faith Exp $";
   struct utsname    uts;
   
   if (buffer) return buffer;
   uname( &uts );
   buffer = xmalloc(256);
   sprintf( buffer,
	    "%s %s (%s %s)", err_program_name(), id_string( id ),
	    uts.sysname,
	    uts.release );
   return buffer;
}

static void banner( void )
{
   fprintf( stderr, "%s\n", pgw_get_banner() );
   fprintf( stderr,
	    "Copyright 1997 Rickard E. Faith (faith@cs.unc.edu)\n" );
}

static void license( void )
{
   static const char *license_msg[] = {
     "",
     "This program is free software; you can redistribute it and/or modify it",
     "under the terms of the GNU General Public License as published by the",
     "Free Software Foundation; either version 1, or (at your option) any",
     "later version.",
     "",
     "This program is distributed in the hope that it will be useful, but",
     "WITHOUT ANY WARRANTY; without even the implied warranty of",
     "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU",
     "General Public License for more details.",
     "",
     "You should have received a copy of the GNU General Public License along",
     "with this program; if not, write to the Free Software Foundation, Inc.,",
     "675 Mass Ave, Cambridge, MA 02139, USA.",
   };
   const char        **p = license_msg;
   
   banner();
   while (*p) fprintf( stderr, "   %s\n", *p++ );
}
    
static void help( void )
{
   static const char *help_msg[] = {
      "-h --help            give this help",
      "-L --license         display software license",
      "-v --verbose         verbose mode",
      "-V --version         display version number",
      "-d --debug <option>  select debug option",
      "-c                   GCIDE",
      0 };
   const char        **p = help_msg;

   banner();
   while (*p) fprintf( stderr, "%s\n", *p++ );
}

int main( int argc, char **argv )
{
   int                c;
   struct option      longopts[]  = {
      { "verbose", 0, 0, 'v' },
      { "version", 0, 0, 'V' },
      { "debug",   1, 0, 'd' },
      { "help",    0, 0, 'h' },
      { "license", 0, 0, 'L' },
      { "gcide",    0, 0, 'c' },
      { 0,         0, 0,  0  }
   };

   maa_init(argv[0]);
   dbg_register(DBG_VERBOSE, "verbose");
   dbg_register(DBG_PARSE,   "parse");
   dbg_register(DBG_SCAN,    "scan");
   dbg_register(DBG_NEWLINE, "newline");

   while ((c = getopt_long( argc, argv,
			    "vVd:hLc", longopts, NULL )) != EOF)
      switch (c) {
      case 'v': dbg_set( "verbose" ); break;
      case 'V': banner(); exit(1);    break;
      case 'd': dbg_set( optarg );    break;
      case 'L': license(); exit(1);   break;
      case 'c': database = 1;         break;
      case 'h':
      default:  help(); exit(1);      break;
      }

   if (dbg_test(DBG_PARSE))     prs_set_debug(1);
   if (dbg_test(DBG_SCAN))      yy_flex_debug = 1;
   else                         yy_flex_debug = 0;

   if (database == 1) fmt_open(dbg_test(DBG_VERBOSE) ? NULL : "gcide");
   else               fmt_open(dbg_test(DBG_VERBOSE) ? NULL : "web1913");
   prs_stream( stdin, "[stdin]" );
   fmt_close();
   return 0;
}
